## Librerias

In [139]:
import pandas as pd

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import style
import seaborn as sns
import urllib.request
from PIL import Image

import urllib.request
from PIL import Image



## Data

In [146]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train_nor = pd.read_csv("C:/Users/javier.sdiaz/Desktop/BootCamp/DS102024/3-MachineLearning/EDA_2/train_normalizado.csv",sep=';')

## Data exploration

In [147]:
train_nor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   laptop_ID      391 non-null    int64  
 1   Company        391 non-null    int64  
 2   TypeName       391 non-null    int64  
 3   Inches         391 non-null    float64
 4   Ram            391 non-null    float64
 5   Memory         391 non-null    float64
 6   Weight         391 non-null    float64
 7   IsTouchscreen  391 non-null    int64  
 8   Memory_Cpu     391 non-null    float64
 9   OpsysType      391 non-null    int64  
 10  Manufacturer   391 non-null    int64  
dtypes: float64(5), int64(6)
memory usage: 33.7 KB


In [142]:
train_nor.head()

laptop_ID  Company  TypeName  Inches   Ram  Memory  Weight  IsTouchscreen  \
0        539        2         3    15.6   8.0   256.0    2.00              0   
1        327        2         3    14.0   8.0   256.0    2.00              0   
2        563        9         3    13.3   4.0    32.0    1.35              0   
3         13        1         4    15.4  16.0   256.0    1.83              0   
4        935        5         4    15.6   8.0   256.0    1.84              0   

   Memory_Cpu  OpsysType  Manufacturer  
0        2.70          5             3  
1        2.70          5             2  
2        1.44          5             2  
3        2.80          6             0  
4        2.50          5             0

## Data processing

In [148]:
train_nor = train_nor[['Company','Ram','Weight','Inches','TypeName','Memory','IsTouchscreen','Memory_Cpu','OpsysType','Manufacturer']]

In [149]:
train_nor

Company   Ram  Weight  Inches  TypeName  Memory  IsTouchscreen  \
0          2   8.0    2.00    15.6         3   256.0              0   
1          2   8.0    2.00    14.0         3   256.0              0   
2          9   4.0    1.35    13.3         3    32.0              0   
3          1  16.0    1.83    15.4         4   256.0              0   
4          5   8.0    1.84    15.6         4   256.0              0   
..       ...   ...     ...     ...       ...     ...            ...   
386        7   8.0    1.44    13.3         3   256.0              0   
387        3   8.0    1.20    13.3         4   256.0              0   
388        7   6.0    2.40    15.6         3   256.0              0   
389        5   8.0    1.43    14.0         3   256.0              0   
390        7   8.0    1.99    15.6         3   512.0              0   

     Memory_Cpu  OpsysType  Manufacturer  
0          2.70          5             3  
1          2.70          5             2  
2          1.44          5             2  
3          2.80          6             0  
4          2.50          5             0  
..          ...        ...           ...  
386        2.50          5             2  
387        1.60          5             2  
388        2.50          5             3  
389        2.30          5             2  
390        2.70          5             2  

[391 rows x 10 columns]

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [145]:
X = train_nor.drop(['Price_euros'], axis=1)
y = train_nor['Price_euros'].copy()
X.shape

KeyError: "['Price_euros'] not found in axis"

In [84]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [85]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [86]:
X_train

Company  Ram  Weight  Inches  TypeName  Memory  IsTouchscreen  \
25        10  4.0    1.70    14.0         3   128.0          False   
84         4  4.0    2.25    15.6         3  1024.0          False   
10        16  8.0    2.00    15.6         3   256.0          False   
342       10  4.0    1.90    15.6         3  1024.0          False   
890        4  8.0    1.29    13.3         4   128.0           True   
..       ...  ...     ...     ...       ...     ...            ...   
106        4  8.0    2.02    15.6         3  2048.0          False   
270        7  8.0    1.90    15.6         3   500.0          False   
860        4  8.0    1.36    14.0         4   256.0          False   
435       10  8.0    2.20    15.6         3   256.0          False   
102        0  8.0    3.30    17.3         3  1024.0          False   

     Memory_Cpu  OpsysType  Manufacturer  
25          2.7          5             1  
84          2.0          2             0  
10          2.5          5             1  
342         1.1          4             1  
890         2.5          5             1  
..          ...        ...           ...  
106         1.6          5             0  
270         2.5          5             0  
860         2.5          5             1  
435         2.0          4             2  
102         2.0          5             2  

[729 rows x 10 columns]

### 3. Asignar el modelo (vacío) a una variable


In [87]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()

regressor.fit(X_train, y_train)

LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de RMSE.

In [88]:
from sklearn.metrics import root_mean_squared_error

predictions = regressor.predict(X_test)

rmse = root_mean_squared_error(y_test, predictions)

In [89]:
print(rmse)

436.925272530964


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [90]:
X_pred = pd.read_csv("test_laptop.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

In [91]:
X_pred.shape

(391, 12)

**¿Por qué da error?** 

In [92]:
test = pd.read_csv("C:/Users/javier.sdiaz/Desktop/BootCamp/DS102024/3-MachineLearning/EDA_2/test.csv")

In [ ]:
test

laptop_ID   Company                 Product   TypeName  Inches  \
0          539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1          327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2          563  Mediacom           SmartBook 130   Notebook    13.3   
3           13     Apple             MacBook Pro  Ultrabook    15.4   
4          935        HP           EliteBook 850  Ultrabook    15.6   
..         ...       ...                     ...        ...     ...   
386        742    Lenovo             ThinkPad 13   Notebook    13.3   
387        660      Dell                  XPS 13  Ultrabook    13.3   
388        983    Lenovo       IdeaPad 310-15IKB   Notebook    15.6   
389       1137        HP          EliteBook 1040   Notebook    14.0   
390        556    Lenovo           ThinkPad T570   Notebook    15.6   

                       ScreenResolution                          Cpu   Ram  \
0       IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                     Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2           IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3    IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                     Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   
..                                  ...                          ...   ...   
386         IPS Panel Full HD 1920x1080   Intel Core i5 7200U 2.5GHz   8GB   
387                   Full HD 1920x1080   Intel Core i5 8250U 1.6GHz   8GB   
388                   Full HD 1920x1080   Intel Core i5 7200U 2.5GHz   6GB   
389                   Full HD 1920x1080   Intel Core i5 6200U 2.3GHz   8GB   
390         IPS Panel Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   

                   Memory                      Gpu       OpSys  Weight  
0    256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1               256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2      32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3               256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4               256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg  
..                    ...                      ...         ...     ...  
386             256GB SSD    Intel HD Graphics 620  Windows 10  1.44kg  
387             256GB SSD   Intel UHD Graphics 620  Windows 10   1.2kg  
388             256GB SSD     Nvidia GeForce 920MX  Windows 10   2.4kg  
389             256GB SSD    Intel HD Graphics 520   Windows 7  1.43kg  
390             512GB SSD    Intel HD Graphics 620  Windows 10  1.99kg  

[391 rows x 12 columns]

In [94]:
import re

def separate_screen_resolution(resolution_str):
    # Extract resolution
    resolution_match = re.search(r'(\d+)\s*x\s*(\d+)', str(resolution_str))
    resolution = f"{resolution_match.group(1)}x{resolution_match.group(2)}" if resolution_match else None
    
    # Extract screen type
    screen_types = ['Touchscreen', 'IPS Panel', 'Full HD', 'Quad HD+', '4K Ultra HD', 'Retina Display']
    found_types = [type for type in screen_types if type.lower() in resolution_str.lower()]
    screen_type = ' / '.join(found_types) if found_types else None
    
    return screen_type, resolution

# Apply the function
test[['ScreenType', 'Resolution']] = test['ScreenResolution'].apply(separate_screen_resolution).apply(pd.Series)

In [95]:
# Crear una nueva columna 'IsTouchscreen' con True si contiene 'Touchscreen' y False si no
test['IsTouchscreen'] = test['ScreenType'].str.contains('Touchscreen', case=False, na=False)

In [ ]:
# Separar el tipo de CPU y la memoria (en GHz)
test[['Cpu_Type', 'Memory_Cpu']] = test['Cpu'].str.extract(r'^(.*)\s(\d+\.?\d*GHz)$')

test.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  \
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg   
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg   
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg   
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg   
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg   

                   ScreenType Resolution  IsTouchscreen             Cpu_Type  \
0     IPS Panel / 4K Ultra HD  3840x2160          False  Intel Core i7 7500U   
1                     Full HD  1920x1080          False  Intel Core i7 7500U   
2         IPS Panel / Full HD  1920x1080          False  Intel Atom x5-Z8350   
3  IPS Panel / Retina Display  2880x1800          False        Intel Core i7   
4                     Full HD  1920x1080          False  Intel Core i7 6500U   

  Memory_Cpu  
0     2.7GHz  
1     2.7GHz  
2    1.44GHz  
3     2.8GHz  
4     2.5GHz

In [97]:
# Función para limpiar los valores y normalizarlos
def clean_and_normalize_memory(value):
    # Asegurarnos de que el valor es un string
    value = str(value).lower().strip()
    
    # Usar una expresión regular para capturar el número y la unidad opcional
    match = re.match(r'(\d+(\.\d+)?)(\s*(gb|mb|tb)?)?', value)
    
    if match:
        # Extraer el número y la unidad
        number = float(match.group(1))  # Captura el número
        unit = match.group(3).strip() if match.group(3) else ''  # Captura la unidad (puede estar vacía)
        
        # Convertir todo a GB
        if 'gb' in unit or unit == '':
            return number  # Ya está en GB o es solo un número
        elif 'mb' in unit:
            return number / 1024  # Convertir de MB a GB
        elif 'tb' in unit:
            return number * 1024  # Convertir de TB a GB
    return None  # Si no es válido, devolver None

# Aplicar la función a la columna Memory
test['Memory'] = test['Memory'].apply(clean_and_normalize_memory)

In [ ]:
test[['OpsysType', 'Version']] = test['OpSys'].str.extract(r'^(Windows|Chrome OS|Linux|No OS|Mac OS X|macOS|Android)(?:\s+(.*))?$')

test.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

   Memory                      Gpu       OpSys  Weight  \
0   256.0  Nvidia GeForce GTX 960M  Windows 10     2kg   
1   256.0    Intel HD Graphics 620  Windows 10     2kg   
2    32.0        Intel HD Graphics  Windows 10  1.35kg   
3   256.0       AMD Radeon Pro 555       macOS  1.83kg   
4   256.0      AMD Radeon R7 M365X  Windows 10  1.84kg   

                   ScreenType Resolution  IsTouchscreen             Cpu_Type  \
0     IPS Panel / 4K Ultra HD  3840x2160          False  Intel Core i7 7500U   
1                     Full HD  1920x1080          False  Intel Core i7 7500U   
2         IPS Panel / Full HD  1920x1080          False  Intel Atom x5-Z8350   
3  IPS Panel / Retina Display  2880x1800          False        Intel Core i7   
4                     Full HD  1920x1080          False  Intel Core i7 6500U   

  Memory_Cpu OpsysType Version  
0     2.7GHz   Windows      10  
1     2.7GHz   Windows      10  
2    1.44GHz   Windows      10  
3     2.8GHz     macOS     NaN  
4     2.5GHz   Windows      10

In [ ]:
test[['Manufacturer', 'MemoryGraphic']] = test['Gpu'].str.extract(r'^(Intel|AMD|Nvidia|ARM)\s+([\w\s]+)')

test.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

   Memory                      Gpu  ...  Weight                  ScreenType  \
0   256.0  Nvidia GeForce GTX 960M  ...     2kg     IPS Panel / 4K Ultra HD   
1   256.0    Intel HD Graphics 620  ...     2kg                     Full HD   
2    32.0        Intel HD Graphics  ...  1.35kg         IPS Panel / Full HD   
3   256.0       AMD Radeon Pro 555  ...  1.83kg  IPS Panel / Retina Display   
4   256.0      AMD Radeon R7 M365X  ...  1.84kg                     Full HD   

  Resolution IsTouchscreen             Cpu_Type Memory_Cpu OpsysType Version  \
0  3840x2160         False  Intel Core i7 7500U     2.7GHz   Windows      10   
1  1920x1080         False  Intel Core i7 7500U     2.7GHz   Windows      10   
2  1920x1080         False  Intel Atom x5-Z8350    1.44GHz   Windows      10   
3  2880x1800         False        Intel Core i7     2.8GHz     macOS     NaN   
4  1920x1080         False  Intel Core i7 6500U     2.5GHz   Windows      10   

  Manufacturer     MemoryGraphic  
0       Nvidia  GeForce GTX 960M  
1        Intel   HD Graphics 620  
2        Intel       HD Graphics  
3          AMD    Radeon Pro 555  
4          AMD   Radeon R7 M365X  

[5 rows x 21 columns]

In [100]:
test['Model Base'] = test['Product'].str.extract(r'^([^(/]+)')  # Extrae el texto antes de un paréntesis o diagonal
test['Specifications'] = test['Product'].str.extract(r'\((.*)\)|/(.*)').bfill(axis=1)[0]  # Extrae el contenido entre paréntesis o después de una diagonal

test.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

   Memory                      Gpu  ... Resolution IsTouchscreen  \
0   256.0  Nvidia GeForce GTX 960M  ...  3840x2160         False   
1   256.0    Intel HD Graphics 620  ...  1920x1080         False   
2    32.0        Intel HD Graphics  ...  1920x1080         False   
3   256.0       AMD Radeon Pro 555  ...  2880x1800         False   
4   256.0      AMD Radeon R7 M365X  ...  1920x1080         False   

              Cpu_Type Memory_Cpu  OpsysType Version Manufacturer  \
0  Intel Core i7 7500U     2.7GHz    Windows      10       Nvidia   
1  Intel Core i7 7500U     2.7GHz    Windows      10        Intel   
2  Intel Atom x5-Z8350    1.44GHz    Windows      10        Intel   
3        Intel Core i7     2.8GHz      macOS     NaN          AMD   
4  Intel Core i7 6500U     2.5GHz    Windows      10          AMD   

      MemoryGraphic              Model Base Specifications  
0  GeForce GTX 960M  Zenbook UX510UW-FI095T            NaN  
1   HD Graphics 620  ZenBook UX410UA-GV183T            NaN  
2       HD Graphics           SmartBook 130            NaN  
3    Radeon Pro 555             MacBook Pro            NaN  
4   Radeon R7 M365X           EliteBook 850            NaN  

[5 rows x 23 columns]

In [101]:
test['Ram'] = test['Ram'].astype(str)
test['Ram'] = test['Ram'].str.replace('GB', '').astype(float)
test['Memory'] = test['Memory'].astype(str)
test['Memory'] = test['Memory'].str.replace('GHz', '').astype(float)
test['Weight']= test['Weight'].astype(str)
test['Weight'] = test['Weight'].str.replace('kg', '').astype(float)
test['Memory_Cpu']= test['Memory_Cpu'].astype(str)
test['Memory_Cpu'] = test['Memory_Cpu'].str.replace('GHz', '').astype(float)

In [ ]:
df_filtered_nor = test.drop(columns=['ScreenResolution', 'Cpu','Gpu','OpSys','Product'])
df_filtered_nor

laptop_ID   Company   TypeName  Inches   Ram  Memory  Weight  \
0          539      Asus   Notebook    15.6   8.0   256.0    2.00   
1          327      Asus   Notebook    14.0   8.0   256.0    2.00   
2          563  Mediacom   Notebook    13.3   4.0    32.0    1.35   
3           13     Apple  Ultrabook    15.4  16.0   256.0    1.83   
4          935        HP  Ultrabook    15.6   8.0   256.0    1.84   
..         ...       ...        ...     ...   ...     ...     ...   
386        742    Lenovo   Notebook    13.3   8.0   256.0    1.44   
387        660      Dell  Ultrabook    13.3   8.0   256.0    1.20   
388        983    Lenovo   Notebook    15.6   6.0   256.0    2.40   
389       1137        HP   Notebook    14.0   8.0   256.0    1.43   
390        556    Lenovo   Notebook    15.6   8.0   512.0    1.99   

                     ScreenType Resolution  IsTouchscreen  \
0       IPS Panel / 4K Ultra HD  3840x2160          False   
1                       Full HD  1920x1080          False   
2           IPS Panel / Full HD  1920x1080          False   
3    IPS Panel / Retina Display  2880x1800          False   
4                       Full HD  1920x1080          False   
..                          ...        ...            ...   
386         IPS Panel / Full HD  1920x1080          False   
387                     Full HD  1920x1080          False   
388                     Full HD  1920x1080          False   
389                     Full HD  1920x1080          False   
390         IPS Panel / Full HD  1920x1080          False   

                Cpu_Type  Memory_Cpu OpsysType Version Manufacturer  \
0    Intel Core i7 7500U        2.70   Windows      10       Nvidia   
1    Intel Core i7 7500U        2.70   Windows      10        Intel   
2    Intel Atom x5-Z8350        1.44   Windows      10        Intel   
3          Intel Core i7        2.80     macOS     NaN          AMD   
4    Intel Core i7 6500U        2.50   Windows      10          AMD   
..                   ...         ...       ...     ...          ...   
386  Intel Core i5 7200U        2.50   Windows      10        Intel   
387  Intel Core i5 8250U        1.60   Windows      10        Intel   
388  Intel Core i5 7200U        2.50   Windows      10       Nvidia   
389  Intel Core i5 6200U        2.30   Windows       7        Intel   
390  Intel Core i7 7500U        2.70   Windows      10        Intel   

        MemoryGraphic              Model Base Specifications  
0    GeForce GTX 960M  Zenbook UX510UW-FI095T            NaN  
1     HD Graphics 620  ZenBook UX410UA-GV183T            NaN  
2         HD Graphics           SmartBook 130            NaN  
3      Radeon Pro 555             MacBook Pro            NaN  
4     Radeon R7 M365X           EliteBook 850            NaN  
..                ...                     ...            ...  
386   HD Graphics 620             ThinkPad 13            NaN  
387  UHD Graphics 620                  XPS 13            NaN  
388     GeForce 920MX       IdeaPad 310-15IKB            NaN  
389   HD Graphics 520          EliteBook 1040            NaN  
390   HD Graphics 620           ThinkPad T570            NaN  

[391 rows x 18 columns]

In [103]:
# Convertir la columna 'laptop_ID' a índice
#df_ = test.set_index('laptop_ID')

In [104]:
# pasar la columna Manufacturer a numerica
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_filtered_nor['Manufacturer'] = encoder.fit_transform(df_filtered_nor['Manufacturer'])

In [105]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_filtered_nor['OpsysType'] = encoder.fit_transform(df_filtered_nor['OpsysType'])


In [106]:
encoder = LabelEncoder()
df_filtered_nor['Company'] = encoder.fit_transform(df_filtered_nor['Company'])

In [107]:
encoder = LabelEncoder()
df_filtered_nor['TypeName'] = encoder.fit_transform(df_filtered_nor['TypeName'])

In [108]:
df_filtered_nor.columns

Index(['laptop_ID', 'Company', 'TypeName', 'Inches', 'Ram', 'Memory', 'Weight',
       'ScreenType', 'Resolution', 'IsTouchscreen', 'Cpu_Type', 'Memory_Cpu',
       'OpsysType', 'Version', 'Manufacturer', 'MemoryGraphic', 'Model Base',
       'Specifications'],
      dtype='object')

In [109]:
df_filtered_nor.head()

laptop_ID  Company  TypeName  Inches   Ram  Memory  Weight  \
0        539        2         3    15.6   8.0   256.0    2.00   
1        327        2         3    14.0   8.0   256.0    2.00   
2        563        9         3    13.3   4.0    32.0    1.35   
3         13        1         4    15.4  16.0   256.0    1.83   
4        935        5         4    15.6   8.0   256.0    1.84   

                   ScreenType Resolution  IsTouchscreen             Cpu_Type  \
0     IPS Panel / 4K Ultra HD  3840x2160          False  Intel Core i7 7500U   
1                     Full HD  1920x1080          False  Intel Core i7 7500U   
2         IPS Panel / Full HD  1920x1080          False  Intel Atom x5-Z8350   
3  IPS Panel / Retina Display  2880x1800          False        Intel Core i7   
4                     Full HD  1920x1080          False  Intel Core i7 6500U   

   Memory_Cpu  OpsysType Version  Manufacturer     MemoryGraphic  \
0        2.70          5      10             3  GeForce GTX 960M   
1        2.70          5      10             2   HD Graphics 620   
2        1.44          5      10             2       HD Graphics   
3        2.80          6     NaN             0    Radeon Pro 555   
4        2.50          5      10             0   Radeon R7 M365X   

               Model Base Specifications  
0  Zenbook UX510UW-FI095T            NaN  
1  ZenBook UX410UA-GV183T            NaN  
2           SmartBook 130            NaN  
3             MacBook Pro            NaN  
4           EliteBook 850            NaN

In [110]:
df_filtered_nor = df_filtered_nor.drop(columns=['Cpu_Type','MemoryGraphic','Specifications','ScreenType','Resolution','Version','Model Base'])

In [111]:
df_filtered_nor['IsTouchscreen'] = df_filtered_nor['IsTouchscreen'].map({True: 1, False: 0})

In [112]:
df_filtered_nor.to_csv('train_normalizado.csv', index=False, sep=';', encoding='utf-8')

In [113]:
df_filtered_nor.head()

laptop_ID  Company  TypeName  Inches   Ram  Memory  Weight  IsTouchscreen  \
0        539        2         3    15.6   8.0   256.0    2.00              0   
1        327        2         3    14.0   8.0   256.0    2.00              0   
2        563        9         3    13.3   4.0    32.0    1.35              0   
3         13        1         4    15.4  16.0   256.0    1.83              0   
4        935        5         4    15.6   8.0   256.0    1.84              0   

   Memory_Cpu  OpsysType  Manufacturer  
0        2.70          5             3  
1        2.70          5             2  
2        1.44          5             2  
3        2.80          6             0  
4        2.50          5             0

In [129]:
# Buscar 'Asus' en todo el DataFrame
asistencia_asus = X_pred.isin(['Asus'])

# Mostrar las filas donde se encuentra 'Asus'
print(X_pred[asistencia_asus.any(axis=1)])  # .any(axis=1) devuelve filas donde al menos una columna es 'Asus'


    Company   Ram  Weight  Inches            TypeName  Memory  IsTouchscreen  \
0      Asus   8.0   2.000    15.6            Notebook   256.0          False   
1      Asus   8.0   2.000    14.0            Notebook   256.0          False   
21     Asus   8.0   1.200    13.3           Ultrabook   256.0          False   
32     Asus   4.0   2.000    15.6            Notebook  1024.0          False   
37     Asus  16.0   2.900    17.3              Gaming   256.0          False   
43     Asus   8.0   2.591    15.6              Gaming   256.0          False   
50     Asus   4.0   2.370    15.6            Notebook  1024.0          False   
55     Asus   2.0   1.650    14.0            Notebook    32.0          False   
56     Asus   8.0   3.000    17.3              Gaming   128.0          False   
57     Asus   8.0   2.690    17.3            Notebook   256.0          False   
58     Asus  16.0   4.300    17.3              Gaming   512.0          False   
72     Asus   8.0   2.000    15.6       

In [117]:
X_pred = test[['Company', 'Ram', 'Weight', 'Inches', 'TypeName', 'Memory',
       'IsTouchscreen', 'Memory_Cpu', 'OpsysType', 'Manufacturer']]

In [118]:
print("Columnas de X_train:", X_train.columns)
print("Columnas de X_pred:", X_pred.columns)

Columnas de X_train: Index(['Company', 'Ram', 'Weight', 'Inches', 'TypeName', 'Memory',
       'IsTouchscreen', 'Memory_Cpu', 'OpsysType', 'Manufacturer'],
      dtype='object')
Columnas de X_pred: Index(['Company', 'Ram', 'Weight', 'Inches', 'TypeName', 'Memory',
       'IsTouchscreen', 'Memory_Cpu', 'OpsysType', 'Manufacturer'],
      dtype='object')


In [119]:
test = regressor.predict(X_pred)
test

ValueError: could not convert string to float: 'Asus'

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [26]:
X_pred = pd.read_csv("test.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

In [27]:
X_pred = X_pred[['Inches']]
X_pred

Inches
0      15.6
1      14.0
2      13.3
3      15.4
4      15.6
..      ...
386    13.3
387    13.3
388    15.6
389    14.0
390    15.6

[391 rows x 1 columns]

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [70]:
predictions_submit = regressor.predict(X_pred)
predictions_submit

array([1023.22488313, 1160.18967221,  789.7386287 , 1829.8395485 ,
       1015.46954046,  634.0520023 , 1761.71210021,  825.73018922,
        979.11637171,  749.68959905, 1038.25085954, 1915.75411478,
        633.08258447,  634.0520023 ,  911.62941889, 1796.61114221,
        733.40664294, 1018.37779396, 1879.60878036,  622.4189883 ,
       1360.90620538, 1181.3350541 , 1712.75649962,  638.89909147,
        796.52455353,  633.08258447, 1025.16371879, 1879.60878036,
       1145.64840471,  916.47650806, 1937.40155485, 1181.3350541 ,
        624.35782397, 1979.06917242, 1025.16371879,  734.66362264,
        911.62941889, 1719.05771554, 1037.76615063, 1129.16830155,
       1012.56128696, 1132.56126396, 1166.73494184, 1051.8711801 ,
       1081.38995313, 1028.07197229,  629.20491314, 3343.12377825,
       1027.10255446,  751.62843472,  642.29205389, 1038.73556846,
        638.89909147,  629.20491314, 1047.46032896,  544.92427139,
        926.17068639,  911.14470997, 1786.91696387, 1140.80131

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [72]:
sample = pd.read_csv("sample_submission.csv")

In [73]:
sample.head()

laptop_ID  Price_euros
0        539          650
1        327          650
2        563          650
3         13          650
4        935          500

In [74]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [75]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [76]:
submission.head()

laptop_ID  Price_euros
0        539  1023.224883
1        327  1160.189672
2        563   789.738629
3         13  1829.839548
4        935  1015.469540

In [77]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [78]:
def chequeator(df_to_submit,version):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit['laptop_ID'].all() == sample['laptop_ID'].all():
                print("You're ready to submit!")
                submission.to_csv(f"submission_{version}.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [79]:
chequeator(submission,1)

You're ready to submit!
